src/Features/Core/Portable/NavigateTo/NavigateToProjectSearcher.cs

```diff
@@ -0,0 +1,347 @@
+                _patternMatcher = new PatternMatcher(pattern, allowFuzzyMatching: true);
+            }
+
+            public void Dispose()
+            {
+                _patternMatcher.Dispose();
+            }
+
+            internal async Task<ImmutableArray<INavigateToSearchResult>> DoAsync()
+            {
+                // Search all documents in parallel.  Each search gets a dedicated 'temp array' 
+                // that it can place its results into without having to worry about locking.  
+                // Once all the tasks are complete, we'll aggregate all the 'temp arrays' into
+                // the final result.
+                var documentSearchTasks = ArrayBuilder<Task>.GetInstance();
+                var tempArrays = ArrayBuilder<ArrayBuilder<INavigateToSearchResult>>.GetInstance();
+
+                try
+                {
+                    // Spawn off a task per document to search.
+                    SpawnDocumentSearchTasks(documentSearchTasks, tempArrays);
+
+                    // Wait for all the tasks to finish.
+                    await Task.WhenAll(documentSearchTasks).ConfigureAwait(false);
+                    _cancellationToken.ThrowIfCancellationRequested();
+
+                    // Collect all the results into the final result.
+                    var result = ArrayBuilder<INavigateToSearchResult>.GetInstance();
+                    foreach (var tempArray in tempArrays)
+                    {
+                        result.AddRange(tempArray);
+                    }
+
+                    return result.ToImmutableAndFree();
+                }
+                finally
+                {
+                    foreach (var tempArray in tempArrays)
+                    {
+                        tempArray.Free();
+                    }
+
+                    tempArrays.Free();
+                    documentSearchTasks.Free();
+                }
+            }
+
+            private void SpawnDocumentSearchTasks(ArrayBuilder<Task> tasks, ArrayBuilder<ArrayBuilder<INavigateToSearchResult>> tempArrays)
+            {
+                foreach (var document in _project.Documents)
+                {
+                    _cancellationToken.ThrowIfCancellationRequested();
+
+                    if (_searchDocument != null && document != _searchDocument)
+                    {
+                        continue;
+                    }
+
+                    var tempArray = ArrayBuilder<INavigateToSearchResult>.GetInstance();
+                    tempArrays.Add(tempArray);
+
+                    // Kick off a fresh task so searching can happen in parallel.
+                    tasks.Add(Task.Run(() => SearchDocumentAsync(document, tempArray)));
+                }
+            }
+
+            private async Task SearchDocumentAsync(
+                Document document, ArrayBuilder<INavigateToSearchResult> results)
+            {
+                _cancellationToken.ThrowIfCancellationRequested();
```

Comment:
:bulb: If you pass the cancellation token to `Task.Run` it's extremely unlikely that you'll hit this line with cancellation requested. The aggregate penalty for simply removing this line will likely be smaller than is measurable.